### PR TITLE
Add initial pipeline resources

### PR DESCRIPTION
**Pull request by janethavi**

### CODE REVIEW

File: main.groovy [hunk 1]
```diff
@@ -0,0 +1,343 @@
+
```
Comment (VimukthiPerera) [inline main.groovy:1]: Please remove the empty line.

Comment (janethavi) [reply main.groovy]: Fixed

File: scripts/apim/deploy.sh [hunk 2]
```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
+# -------------------------------------------------------------------------------------
+#
+# Copyright (c) 2022, WSO2 Inc. (http://www.wso2.com). All Rights Reserved.
```
Comment (VimukthiPerera) [inline scripts/apim/deploy.sh:4]: Shall we change the license header to the open source licensing?

**Review by VimukthiPerera — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -1,33 +1,37 @@
...
+9. **product_repository**- The product repo where the test scripts are existing.
+Allowed Value Format- <https://github.com/janethavi/product-apim.git>
+10. **product_test_branch**- The repo branch where the test script is existing.
+Allowed Values- product-scenarios-3.2.0,
+11. **product_test_script**- The location of the test script.
+Allowed Value Format- product-scenarios/test.sh
+
+## step by step guide to on how to add this repo to a new pipeline
+
+Check the TestGrid Pipeline Onboarding docuemnt
```
[the remark this answers is not in the document]
Comment (VimukthiPerera) [reply README.md:37]: Spellings mistake: document

File: README.md [hunk 4]
```diff
@@ -1,33 +1,37 @@
...
+## Testgrid Jenkins Pipeline Variables
+
+1. **product**- The WSO2 product that needs to be tested from TestGrid.
+Allowed Values- apim, is
+2. **cfn_repo_url**- The WSO2 product CFN repo name.
+Allowed Value Format- <https://github.com/janethavi/aws-apim.git>
+3. **product_version**- The product version that needs to be tested using testgrid.
+Allowed Value Format- 3.2.0
+4. **product_deployment_region**- The region where the product stack is getting deployed.
+Allowed Values- us-east-2,
```
Comment (VimukthiPerera) [inline README.md:19]: Please remove the comma at the end of line or add other regions of there are any.

File: main.groovy [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+* specific language governing permissions and limitations
+* under the License.
+*
+*/
+import groovy.io.FileType
+import hudson.model.*
+
+def deploymentDirectories = []
+def updateType = ""
+def s3BucketName = "janeth-tg-logs"
```
[the remark this answers is not in the document]
Comment (janethavi) [reply main.groovy:24]: Shall I create a new separate S3 bucket to contain all the logs?

File: README.md [hunk 6]
```diff
@@ -1,33 +1,37 @@
...
+1. **product**- The WSO2 product that needs to be tested from TestGrid.
+Allowed Values- apim, is
+2. **cfn_repo_url**- The WSO2 product CFN repo name.
+Allowed Value Format- <https://github.com/janethavi/aws-apim.git>
+3. **product_version**- The product version that needs to be tested using testgrid.
+Allowed Value Format- 3.2.0
+4. **product_deployment_region**- The region where the product stack is getting deployed.
+Allowed Values- us-east-2, 
+5. **os_list**- The OS and its version. If there are multiple parameters, please add them by separating them by a ","(Comma).
+Allowed Values- CentOS7,
```
Comment (VimukthiPerera) [inline README.md:21]: Please add the rest of the values for this and the parameters below.

File: main.groovy [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+def s3BuildLogPath = ""
+
+pipeline {
+agent {label 'pipeline-slave'}
```
Comment (VimukthiPerera) [inline main.groovy:28]: Shall we rename the label to `pipeline-agent`?

File: main.groovy [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+                if (use_wum.toBoolean()){
+                    aws_repo_branch="${product_version}-new"
+                    updateType="wum"
+                }else{
+                    aws_repo_branch="${product_version}-u2-new"
+                    updateType="u2"
+                }
+                dir("aws-"+product) {
+                    git branch: "${aws_repo_branch}",
+                    credentialsId: "JANETH_GITHUB_TEMP",
```
Comment (VimukthiPerera) [inline main.groovy:111]: Shall we configure new Credentials(wso2 common user) for a GitHub user and use the ID here.

File: main.groovy [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+                        ./scripts/write-parameter-file.sh "WUMUsername" ${wumUserName} "${WORKSPACE}/parameters/parameters.json"
+                        echo "Writting DB password to parameter file"
+                        ./scripts/write-parameter-file.sh "DBPassword" ${dbPassword} "${WORKSPACE}/parameters/parameters.json"
+                        echo "Writting S3 access key id to parameter file"
+                        ./scripts/write-parameter-file.sh "S3AccessKeyID" ${s3accessKey} "${WORKSPACE}/parameters/parameters.json"
+                        echo "Writting S3 secret access key to parameter file"
+                        ./scripts/write-parameter-file.sh "S3SecretAccessKey" ${s3secretKey} "${WORKSPACE}/parameters/parameters.json"
+                    '''
+                }
+                withCredentials([usernamePassword(credentialsId: 'JANETH_GITHUB_TEMP', usernameVariable: 'githubUserName', passwordVariable: 'githubPassword')])
```
Comment (VimukthiPerera) [inline main.groovy:145]: Replace credential ID here as well

File: main.groovy [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+                String[] cloudformationLocation = []
+                switch(product) {
+                    case "apim":
+                        cloudformationLocation = ["${WORKSPACE}/aws-apim/apim/Minimum-HA/apim.yaml"]
+                        break;
+                    case "is":
+                        // The deployment is done in the indexed order
+                        cloudformationLocation = ["${WORKSPACE}/aws-is/is/Minimum-HA/identity.yaml", "${WORKSPACE}/aws-is/is-samples/test-is-samples.yml"]
+                        break;
+                    default:
```
Comment (VimukthiPerera) [inline main.groovy:233]: Shall we include, ob, mi, ei, and esb as well?

File: main.groovy [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+            <b>Build Info:</b>
+            <small><a href="${BUILD_URL}">${BUILD_URL}</a></small>
+            </font>
+        </p>
+        <br/>
+        <em>Tested by WSO2 Jenkins TestGrid Pipeline.</em>
+        </div>
+        """
+    subject="[TestGrid][${updateType.toUpperCase()}][${product.toUpperCase()}:${product_version}][SCE]-Build ${currentBuild.currentResult}-#${env.BUILD_NUMBER}"
+    emailext(to: "janeth@wso2.com",
```
Comment (VimukthiPerera) [inline main.groovy:346]: Change email to testgrid group

Comment (janethavi) [reply main.groovy]: Added `testgrid@wso2.com`

File: main.groovy (hunk 9 above)
Comment (janethavi) [reply main.groovy]: Fixed by replacing to the new id `WSO2_GITHUB_TOKEN`

File: main.groovy (hunk 8 above)
Comment (janethavi) [reply main.groovy]: Fixed by replacing to the new id `WSO2_GITHUB_TOKEN`

File: main.groovy (hunk 7 above)
Comment (janethavi) [reply main.groovy]: Fixed

File: README.md (hunk 6 above)
Comment (janethavi) [reply README.md]: Fixed

File: scripts/apim/deploy.sh (hunk 2 above)
Comment (janethavi) [reply scripts/apim/deploy.sh]: Added the Apache licence to shell scripts

File: main.groovy [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+                        ),
+                        string(
+                            name: 'product_test_script', 
+                            defaultValue: '',
+                            description: 'The location of the test script. Check the README file on https://github.com/wso2/testgrid-jenkins-library/tree/main/README.md',
+                            trim: false
+                        )
+                    ])
+                ])
+                aws_repo_branch="" //need to exit if the branch name is invalid
```
[the remark this answers is not in the document]
Comment (janethavi) [reply main.groovy]: Cloning the repo is done by a plugin. I checked if there is a way of handling errors (when the branch is wrong or any other errors)but looks like the pipeline will end while printing an error message.

File: main.groovy (hunk 5 above)
[the remark this answers is not in the document]
Comment (janethavi) [reply main.groovy]: Created a new S3 bucket named `"testgrid-pipeline-logs` and added it to the script

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (janethavi) [reply README.md]: Fixed both the mentioned issues

File: README.md [hunk 13]
```diff
@@ -1,33 +1,37 @@
-# Testgrid Jenkins Shared Library
+# Testgrid Jenkins Library
 
-Library that hosts Testgrid shared components and libraries. This contain a bunch of Jenkins pipeline critical for testgrid work:
+This repo contains all the pipeline code that is being used in the Jenkins TestGrid Pipeline.
```
[the remark this answers is not in the document]
Comment (janethavi) [reply README.md]: Removed the `all the` phrases

File: scripts/apim/deploy.sh [hunk 14]
```diff
@@ -0,0 +1,86 @@
+#!/bin/bash
+# -------------------------------------------------------------------------------------
+#
```
Comment (VimukthiPerera) [inline scripts/apim/deploy.sh:3]: Please add the copyright notice.

Comment (janethavi) [reply scripts/apim/deploy.sh:4]: Fixed

**Review by VimukthiPerera — APPROVED** [review]

File: README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (chamindi-a) [reply README.md]: ack will check the https://github.com/wso2/testgrid-jenkins-library/pull/185/files/a88bb655b0b95d692db8339a49f73c65ef7f50b9#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5

**Pull request merged**